### PR TITLE
Multiple Control Statements in DoConcurrentLoop

### PR DESCRIPTION
**Pull request by adit4443ya**
Towards [Issue](https://github.com/lfortran/lfortran/issues/4900)
To implement multiple control statements in the DoConcurrent loop, we need to first modify the grammar to:  
`| DoConcurrentLoop(do_loop_head* head, expr* shared, expr* local, reduction_expr* reduction, stmt* body)`  
Next, update all references to `make_DoConcurrentLoop_t` as suggested by @Pranavchiku. We will retain `do_loop_head_t* m_head` in the `DoConcurrentLoop_t` struct and the `make_DoConcurrentLoop_t` definition. However, updating the references in `asr.h` causes the type to revert, leading to build errors in the following function:  
- `ASR::asr_t* duplicate_DoConcurrentLoop(DoConcurrentLoop_t* x)`

### Current Changes Made:
- Updated **ASR.asdl** to:  
  `| DoConcurrentLoop(do_loop_head* head, expr* shared, expr* local, reduction_expr* reduction, stmt* body)`

### Primary Steps to Consider:
- Modify `asdl_cpp.py` to correctly generate `asr.h`
- Update `asr.h` and `ast_body_visitor` wherever `make_DoConcurrentLoop_t` is referenced

Before proceeding with the first step, we should consider two options:
- Create a vector of `do_loop_head_t`
- Modify `do_loop_head_t` to function like a list

### CODE REVIEW

**Comment by certik** [comment]
I think you just need to remove that one "*" from the generated C++ for the duplicator visitor for `do_loop_head_t`.

**Comment by Pranavchiku** [comment]
So, let's break it down. The problem you have is inside `duplicate_DoConcurrentLoop` which is part of class `BaseExprStmtDuplicator` and we know that it is generated from `asdl_cpp.py`  thus on searching it inside `asdl_cpp.py` it takes us at line: 928. If you go down and look at then `visit_Field` function it has a handling for `do_loop_head` at line `1146` try to add a print statement after line `1143` and see if it comes there for `do_loop_head` it won't as now you are defining it as a sequence ( check generated DoConcurrentLoop_t node ). Hence, it comes inside `field.seq` at line `1091` and that is when you see a handling for `dimension` on line `1096`, similarly apply the diff and you are good to go:

```diff
diff --git a/src/libasr/asdl_cpp.py b/src/libasr/asdl_cpp.py
index 0907f7a8d..ce4808375 100644
--- a/src/libasr/asdl_cpp.py
+++ b/src/libasr/asdl_cpp.py
@@ -1093,7 +1093,8 @@ class ExprStmtDuplicatorVisitor(ASDLVisitor):
                 if (field.type != "call_arg" and
                     field.type != "array_index" and
                     field.type != "alloc_arg" and
-                    field.type != "dimension"):
+                    field.type != "dimension" and
+                    field.type != "do_loop_head"):
                     pointer_char = '*'
                 self.emit("Vec<%s_t%s> m_%s;" % (field.type, pointer_char, field.name), level)
                 self.emit("m_%s.reserve(al, x->n_%s);" % (field.name, field.name), level)
```

**Comment by Pranavchiku** [comment]
It is easy once you find out the code flow and the only thing you guys need to develop is thought process on how to find it. This is how I do, more verbose in the comment.

**Comment by certik** [comment]
Thanks @Pranavchiku for the details how to do this workflow. @adit4443ya let us know if this is good enough to unblock you.

Btw, it would be better to refactor the code above to handle all `(...)` ASDL nodes correctly (if that is the issue), but that's a separate task to this PR.

**Comment by adit4443ya** [comment]
Changes

1. **Changes in `asdl_cpp.py`**: 
   - Modifications were made in the respective visitor classes:
     - `ExprStmtDuplicatorVisitor`
     - `DeserializationVisitorVisitor`
     
2. **Functions Updated**: 
   - Two functions were using `make_DoConcurrentLoop_t`: 
     - `ASR::asr_t* duplicate_DoConcurrentLoop`
     - `asr_t* deserialize_DoLoop()`



3. **Struct Update**: 
   - Changed `ASNodeVisitor::visitConstructor` in `asdl_cpp.py` to include:
     - `Vec<do_loop_head_t> m_head;`
   - This leads to` Vec<do_loop_heaad_t> m_head` in the `make_DoConcurrentLoop_t` and struct `DoConcurrentLoop`


 **Remaining Work**: 
   - The remaining task involves handling multiple `do_loop_head_t` in `openmp.cpp` within `DoConcurrentVisitor->create_lcompilers_function`.

Once this final change is implemented,  we can expect a successful build as far as I can see it.

**Comment by Pranavchiku** [comment]
For now do not handle multiple control statements in openmp, just make it work for 1st element. Get this PR in.

File: src/libasr/asdl_cpp.py [hunk 1]
```diff
@@ -237,20 +237,40 @@ def visitConstructor(self, cons, base, extra_attributes):
...
+                else:
+                    seq = ""
+                self.emit("%s m_%s;%s" % (type_, f.name, seq), 2)
+                args.append("%s a_%s" % (type_, f.name))
+                lines.append("n->m_%s = a_%s;" % (f.name, f.name))
+                if f.name in ["global_scope", "symtab"]:
+                    lines.append("a_%s->asr_owner = (asr_t*)n;" % (f.name))
+                if f.seq:
+                    args.append("size_t n_%s" % (f.name))
+                    lines.append("n->n_%s = n_%s;" % (f.name, f.name))
```
Comment (certik) [inline src/libasr/asdl_cpp.py:273]: First get everything passing. Then we need to refactor this, there is a lot of code duplication.

File: src/libasr/pass/for_all.cpp [hunk 2]
```diff
@@ -32,9 +32,10 @@ class ForAllVisitor : public ASR::StatementWalkVisitor<ForAllVisitor>
         Vec<ASR::stmt_t*> body;
         body.reserve(al, 1);
         body.push_back(al, assign_stmt);
-
+        Vec<ASR::do_loop_head_t> heads;  // Create a vector of loop heads
+        heads.push_back(al, x.m_head);
```
Comment (Pranavchiku) [inline src/libasr/pass/for_all.cpp:36]:
```suggestion
        Vec<ASR::do_loop_head_t> heads;  // Create a vector of loop heads
        heads.reserve(al, 1);
        heads.push_back(al, x.m_head);
```

**Comment by certik** [comment]
@adit4443ya let's get the CI to pass.

**Comment by Pranavchiku** [comment]
Yes, just update references and CI shall be good.

```
rm -rf ./tests/reference/*
./run_tests.py -u
```

File: tests/reference/asr_openmp-openmp1-3056a3e.stderr [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 6677, in LCompilers::ASR::BaseWalkVisitor<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>::visit_do_loop_head(LCompilers::ASR::do_loop_head_t const&)[0m
+    self().visit_expr(*x.m_v);
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 5228, in LCompilers::ASR::BaseVisitor<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>::visit_expr(LCompilers::ASR::expr_t const&)[0m
+    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 4950, in void LCompilers::ASR::visit_expr_t<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>(LCompilers::ASR::expr_t const&, LCompilers::LFortran::BodyVisitor&)[0m
+    LCOMPILERS_ASSERT(x.base.type == asrType::expr)
+[2m  Binary file "[0m[1m[35m/lib/x86_64-linux-gnu/libc.so.6[39m[0m[2m", in <invalid>[0m
+Segfault: Signal SIGSEGV (segmentation fault) received
```
Comment (certik) [inline tests/reference/asr_openmp-openmp1-3056a3e.stderr:48]: You got an assert error here, so this needs to be fixed.

File: tests/reference/asr_openmp-pragma1-25cfcb2.stderr [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    self().visit_do_loop_head(x.m_head[i]);
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 6679, in LCompilers::ASR::BaseWalkVisitor<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>::visit_do_loop_head(LCompilers::ASR::do_loop_head_t const&)[0m
+    self().visit_expr(*x.m_start);
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 5228, in LCompilers::ASR::BaseVisitor<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>::visit_expr(LCompilers::ASR::expr_t const&)[0m
+    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr.h[39m[0m[2m", line 4950, in void LCompilers::ASR::visit_expr_t<LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::create_and_replace_structType()::StructTypeVisitor>(LCompilers::ASR::expr_t const&, LCompilers::LFortran::BodyVisitor&)[0m
+    LCOMPILERS_ASSERT(x.base.type == asrType::expr)
```
Comment (certik) [inline tests/reference/asr_openmp-pragma1-25cfcb2.stderr:46]: Same here.

File: a
Comment (certik) [inline a:1]: Please remove the binary.

We need to be extremely careful in not committing any binaries into git.

Comment (certik) [reply a:1]: If you remove it with a commit, we must squash all the commits too.

File: src/libasr/asdl_cpp.py [hunk 5]
```diff
@@ -243,8 +243,16 @@ def visitConstructor(self, cons, base, extra_attributes):
...
+            if cons.name=="DoConcurrentLoop":
+                if f.type == "do_loop_head":
+                    self.emit("Vec<%s_t> m_%s;%s" % (f.type, f.name, seq), 2)
+                    args.append("Vec<%s_t>& a_%s" % (f.type, f.name))
+                else:
+                    self.emit("%s m_%s;%s" % (type_, f.name, seq), 2)
+                    args.append("%s a_%s" % (type_, f.name))
+            else:
+                self.emit("%s m_%s;%s" % (type_, f.name, seq), 2)
+                args.append("%s a_%s" % (type_, f.name))
```
Comment (certik) [inline src/libasr/asdl_cpp.py:255]: Does it mean the `convert_type` is implemented incorrectly --- it should return `Vec<do_loop_head_t>`, but it returns `do_loop_head_t`?

File: src/libasr/codegen/asr_to_cpp.cpp [hunk 6]
```diff
@@ -663,11 +663,11 @@ Kokkos::View<T*> from_std_vector(const std::vector<T> &v)
         std::string indent(indentation_level*indentation_spaces, ' ');
         std::string out = indent + "Kokkos::parallel_for(";
         out += "Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(";
-        visit_expr(*x.m_head.m_start);
```
Comment (certik) [inline src/libasr/codegen/asr_to_cpp.cpp:666]: Add assert here that `n_head == 1`.

File: src/libasr/pass/do_loops.cpp [hunk 7]
```diff
@@ -45,7 +45,7 @@ class DoLoopVisitor : public ASR::StatementWalkVisitor<DoLoopVisitor>
     }
 
     void visit_DoConcurrentLoop(const ASR::DoConcurrentLoop_t &x) {
-        ASR::asr_t* do_loop = ASR::make_DoLoop_t(al, x.base.base.loc, s2c(al, ""), x.m_head, x.m_body, x.n_body, nullptr, 0);
+        ASR::asr_t* do_loop = ASR::make_DoLoop_t(al, x.base.base.loc, s2c(al, ""), x.m_head[0], x.m_body, x.n_body, nullptr, 0);
```
Comment (certik) [inline src/libasr/pass/do_loops.cpp:48]: Add assert here that `n_head == 1`, since we do not handle > 1.

File: src/libasr/codegen/asr_to_julia.cpp [hunk 8]
```diff
@@ -1237,7 +1237,7 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
 
     void visit_DoConcurrentLoop(const ASR::DoConcurrentLoop_t& x)
     {
-        const ASR::DoLoop_t do_loop = ASR::DoLoop_t{ x.base, nullptr, x.m_head, x.m_body, x.n_body, nullptr, 0 };
+        const ASR::DoLoop_t do_loop = ASR::DoLoop_t{ x.base, nullptr, x.m_head[0], x.m_body, x.n_body, nullptr, 0 };
```
Comment (certik) [inline src/libasr/codegen/asr_to_julia.cpp:1240]: assert n_head is 1.

**Comment by certik** [comment]
Now when the tests pass, only keep adding commits, don't rebase.

At the end, we will rebase or squash.

File: src/libasr/asdl_cpp.py (hunk 5 above)
Comment (certik) [reply src/libasr/asdl_cpp.py:255]: I looked into it, the type should be `do_loop_head_t*`, which presumably `convert_type` returns.

However this change is incorrect --- it creates `Vec` inside our struct ASR definitions. We should not use `Vec` in there at all. There we must use raw pointers for performance.

Rather, you can use `Vec` when you construct it as a user, if you want to.

Comment (certik) [reply src/libasr/asdl_cpp.py:255]: So the original code I think was correct here, so you should revert it.

Comment (Pranavchiku) [inline src/libasr/asdl_cpp.py:255]: I don't think we need this change.

Comment (Pranavchiku) [reply src/libasr/asdl_cpp.py:255]: Yes, I agree, we don't need this.

File: src/libasr/pass/openmp.cpp [hunk 9]
```diff
@@ -793,7 +793,7 @@ class DoConcurrentVisitor :
             v.current_expr = nullptr;
             v.visit_DoConcurrentLoop(do_loop);
 
-            ASR::do_loop_head_t loop_head = do_loop.m_head;
+            ASR::do_loop_head_t loop_head = do_loop.m_head[0];
```
[the remark this answers is not in the document]
Comment (Pranavchiku) [reply src/libasr/pass/openmp.cpp:796]: just address these and this will be good

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 10]
```diff
@@ -6,7 +6,6 @@
 #include <string>
 #include <cmath>
 #include <set>
-
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:9]: You can keep this empty line here, it separates the C++ imports from LFortran imports.

In general, if you want to do some whitespace or other rearrangement changes, submit a separate PR for that and we can discuss it there.. In this PR let's just focus on do concurrent.

**Review by certik — APPROVED** [review]
I think that this is good now.

This must be squashed, since there is the binary in the commit history.

Just left one last minor comment with the extra empty line.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 11]
```diff
@@ -6,7 +6,6 @@
 #include <string>
 #include <cmath>
 #include <set>
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_body_visitor.cpp:8]:
```suggestion
#include <set>

```

**Review by Pranavchiku — APPROVED** [review]
I addressed the line issue and enabling auto-merge ( squash ) on this PR. This is awesome work! Really appreciate it.

**Comment by certik** [comment]
I agree, great work! Thank you both.
